### PR TITLE
20354 money request participants selector list refactoring

### PR DESCRIPTION
**Pull request by lukemorawski**
### Details

This PR refactors the `MoneyRequestParticipantSelector` and `MoneyRequestParticipantPage` to use new `<SelectionList />` component.
Some other changes had been made, like moving some logic to `useMemo` hooks instead of invalid `useEffect->useState` pattern.
A new param has been added to `<SelectionList />` to allow it to display a component on the right hand side of each list item. This param allows to pass a react node or a function that will render a component.

### Fixed Issues

$ https://github.com/Expensify/App/issues/20354 
PROPOSAL: no proposal (draft pr)

### Tests

- same as QA steps

- [ ] Verify that no errors appear in the JS console

### Offline tests

- same as QA steps

### QA Steps

- [ ] Tap FAB on home page and select "Request Money"
- [ ] Select manual, fill in the amount and press "Next"
- [ ] A list of users/workspaces will appear
- [ ] Tap on list item to create an IOU
- [ ] Or tap on "Split" button on the list item to create a "split"
- [ ] add more items from the list
- [ ] verify that everything works like in the production app
- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

native android

</details>

<details>
<summary>Android: mWeb Chrome</summary>

web android

</details>

<details>
<summary>iOS: Native</summary>

native ios

</details>

<details>
<summary>iOS: mWeb Safari</summary>

web ios

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

web desktop

</details>

<details>
<summary>MacOS: Desktop</summary>

native desktop

</details>

### CODE REVIEW

File: src/pages/iou/steps/MoneyRequstParticipantsPage/MoneyRequestParticipantsSelector.js [hunk 1]
```diff
@@ -287,77 +297,107 @@ function MoneyRequestParticipantsSelector({
...
+                    <SelectCircle isChecked={item.isSelected} />
+                </PressableWithFeedback>
+            ) : (
+                <Button
+                    onPress={() => addParticipantToSelection(item)}
+                    style={[styles.pl2]}
+                    text={translate('iou.split')}
+                    small
+                />
+            );
```
Comment (VickyStash) [inline src/pages/iou/steps/MoneyRequstParticipantsPage/MoneyRequestParticipantsSelector.js:360]: This way it should be more readable
```suggestion
            if (item.isSelected) {
                return (
                    <PressableWithFeedback
                        onPress={() => addParticipantToSelection(item)}
                        disabled={item.isDisabled}
                        role={CONST.ACCESSIBILITY_ROLE.CHECKBOX}
                        accessibilityLabel={CONST.ACCESSIBILITY_ROLE.CHECKBOX}
                        style={[styles.flexRow, styles.alignItemsCenter, styles.ml3]}
                    >
                        <SelectCircle isChecked={item.isSelected} />
                    </PressableWithFeedback>
                );
            }

            return (
                <Button
                    onPress={() => addParticipantToSelection(item)}
                    style={[styles.pl2]}
                    text={translate('iou.split')}
                    small
                />
            );
```

Comment (lukemorawski) [reply src/pages/iou/steps/MoneyRequstParticipantsPage/MoneyRequestParticipantsSelector.js:360]: I concur!

File: src/components/SelectionList/BaseListItem.js [hunk 2]
```diff
@@ -33,6 +34,14 @@ function BaseListItem({
     const isUserItem = lodashGet(item, 'icons.length', 0) > 0;
     const ListItem = isUserItem ? UserListItem : RadioListItem;
 
+    const rightHandSideComponentRender = () => {
+        if (canSelectMultiple || !rightHandSideComponent) {
+            return null; // no right hand side component
+        }
+
+        return typeof rightHandSideComponent === 'function' ? rightHandSideComponent(item) : rightHandSideComponent;
```
Comment (rezkiy37) [inline src/components/SelectionList/BaseListItem.js:42]: Let's use the `if` block here.
```suggestion
        if (typeof rightHandSideComponent === 'function' ) {
            return rightHandSideComponent(item)
        }
        
        return rightHandSideComponent;
```

Comment (lukemorawski) [reply src/components/SelectionList/BaseListItem.js:42]: I'm doing it only for you!

File: src/components/SelectionList/BaseListItem.js [hunk 3]
```diff
@@ -33,6 +34,18 @@ function BaseListItem({
     const isUserItem = lodashGet(item, 'icons.length', 0) > 0;
     const ListItem = isUserItem ? UserListItem : RadioListItem;
 
+    const rightHandSideComponentRender = () => {
+        if (canSelectMultiple || !rightHandSideComponent) {
+            return null; // no right hand side component
```
Comment (rushatgabhane) [inline src/components/SelectionList/BaseListItem.js:39]: i don't think this comment is needed. the code is self explanatory :)
```suggestion
            return null;
```

**Comment by rushatgabhane** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Review by rushatgabhane — CHANGES_REQUESTED** [review]
@lukemorawski when unselecting an item from list. It appears twice.
In recents and the selected list.



https://github.com/Expensify/App/assets/29673073/b488b674-6ddc-4bf1-b283-db2666bf25af

File: src/components/SelectionList/BaseListItem.js (hunk 3 above)
Comment (lukemorawski) [reply src/components/SelectionList/BaseListItem.js:39]: Looks like another job of Captain Obvious!

**Comment by rushatgabhane** [comment]
@lukemorawski were you able to repro this issue? https://github.com/Expensify/App/pull/32692#pullrequestreview-1774241159

**Comment by lukemorawski** [comment]
@rushatgabhane yep, working on that, but there seems to be some weird changes along the way. So now on `main` money request participants list step is served from `MoneyTemporaryForRefactorRequestParticipantsSelector` instead from `MoneyRequestParticipantsSelector`. They are mostly the samy, but there are some noticable differences. I decided to continue with `MoneyRequestParticipantsSelector` and bring the changes from the new file.

**Comment by lukemorawski** [comment]
@rushatgabhane ok, moved changes to that new component and fixed the problem. The question is what to do with those unused files: `MoneyRequestParticipantsPage.js` and `MoneyRequestParticipantsSelector.js`?

**Comment by rushatgabhane** [comment]
> unused files: MoneyRequestParticipantsPage.js and MoneyRequestParticipantsSelector.js?

@lukemorawski It's used in `IOURequestStepParticipants`, right?

**Review by rushatgabhane — CHANGES_REQUESTED** [review]
Bug: app crash on money request

1. Merge latest `main`
2. Request money
3. Enter amount
4. Go next
5. App crash Cannot find module '@styles/themes/useTheme

image

**Comment by lukemorawski** [comment]
@rushatgabhane merged and fixed imports

**Comment by lukemorawski** [comment]
@rushatgabhane fixed!

**Comment by rushatgabhane** [comment]
@lukemorawski we have the same bug on send money page
1. Click `+`
2. Send money
3. Next
image

**Review by rushatgabhane — CHANGES_REQUESTED** [review]
Bug: selected split gets unselected
1. Request money
2. Split bill with one participant `A`
3. Click Add to split
4. Go back

Expected: In participant selection page, `A` is selected
Actual: No participants are selected

https://github.com/Expensify/App/assets/29673073/2df37d1a-7952-48c0-b209-7b4e6f3911e8

**Comment by lukemorawski** [comment]
> @lukemorawski we have the same bug on send money page
> 
> 1. Click `+`
> 2. Send money
> 3. Next
> 
> image

fixed that

**Comment by rushatgabhane** [comment]
> That's how the app currently behaves on main (before this refactor

yes you're right!

**Comment by rushatgabhane** [comment]
Bug: 
1. Request money
2. In selector search, input some user
3. Click `split`

Expected: The user is selected
image

Actual: The user isn't shown as selected

https://github.com/Expensify/App/assets/29673073/523288fb-c8d9-48f1-8155-abab24741d2f

**Comment by lukemorawski** [comment]
> > That's how the app currently behaves on main (before this refactor
> 
> yes you're right!

So it's good? :) Or should I try to fix that?

**Comment by lukemorawski** [comment]
> Bug:
> 
> 1. Request money
> 2. In selector search, input some user
> 3. Click `split`
> 
> Expected: The user is selected image
> 
> Actual: The user isn't shown as selected
> 
>  Screen.Recording.2024-01-08.at.18.11.26.mov

Fixed that!

**Review by rushatgabhane — APPROVED** [review]
LGTM

**Comment by lukemorawski** [comment]
@rushatgabhane fixed conflicts

**Review by rushatgabhane — APPROVED** [review]

**Comment by cristipaval** [comment]
@lukemorawski and @rushatgabhane , Could you please make sure that the author and reviewer checklists pass?

**Review by rushatgabhane — APPROVED** [review]

**Review by cristipaval — APPROVED** [review]

**Comment by youssef-lr** [comment]
hey @lukemorawski - it seems like `onChangeText` is not debounced in `SectionsList`, can we get this fixed please before this become a deployer blocker?

https://github.com/Expensify/App/assets/9680864/3dd867be-32c7-4329-bdcc-d9889f721cdf

**Comment by lukemorawski** [comment]
@youssef-lr on it!

**Comment by lukemorawski** [comment]
@youssef-lr can you create an ticket from this comment, or should I just create a new PR and mention it here?

**Comment by lukemorawski** [comment]
@youssef-lr here's the quick fix PR
https://github.com/Expensify/App/pull/34443

File: src/pages/iou/request/MoneyTemporaryForRefactorRequestParticipantsSelector.js [hunk 4]
```diff
@@ -213,48 +226,17 @@ function MoneyTemporaryForRefactorRequestParticipantsSelector({
...
+    const headerMessage = useMemo(
+        () =>
+            OptionsListUtils.getHeaderMessage(
+                newChatOptions.personalDetails.length + newChatOptions.recentReports.length !== 0,
+                Boolean(newChatOptions.userToInvite),
+                searchTerm.trim(),
+                maxParticipantsReached,
+                _.some(participants, (participant) => participant.searchText.toLowerCase().includes(searchTerm.trim().toLowerCase())),
+            ),
+        [maxParticipantsReached, newChatOptions.personalDetails.length, newChatOptions.recentReports.length, newChatOptions.userToInvite, participants, searchTerm],
```
[the remark this answers is not in the document]
Comment (lukemorawski) [reply src/pages/iou/request/MoneyTemporaryForRefactorRequestParticipantsSelector.js:238]: any steps to reproduce the crash? cause I had no crashes at all

[the remark this answers is not in the document]
Comment (lukemorawski) [reply src/pages/iou/request/MoneyTemporaryForRefactorRequestParticipantsSelector.js:238]: yep, got it. PR with a fix is on it's way. Should I just mention it here and tag you in the comment, or will there be a new ticket for it?

**Comment by lukemorawski** [comment]
@parasharrajat here's the quick fix PR https://github.com/Expensify/App/pull/34587

**Comment by rushatgabhane** [comment]
Thank you @lukemorawski @parasharrajat 
I'll review it

**Pull request merged**